oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,348 @@
+  and uninstalling packages as needed.
+
+Open edX packages use an ``upgrade`` make target to use ``pip-compile`` to
+automatically update the detailed requirements files (``requirements/*.txt``)
+to use the newest available packages which satisfy the constraints in the
+direct dependencies files.  These generated files are then used anywhere that
+runs a command to install dependencies: ``tox.ini``, ``.travis.yml``, the
+``requirements`` make target (for updating a local development environment),
+etc.
+
+Sometimes ``pip-compile`` will be unable to find a suitable version of a
+dependency for the output file because there are incompatible version
+constraints in the input files and/or the stated installation requirements
+of the other dependencies.  In cases like this, installing and running
+`pipdeptree`_ can help identify the conflicting constraints so at least one
+of them can be sufficiently relaxed such that a version of the dependency
+exists which satisfies them all.
+
+.. _pip-tools: https://github.com/jazzband/pip-tools
+.. _pipdeptree: https://github.com/naiquevin/pipdeptree
+
+Automate Updates of Exact Dependency Specifications
+---------------------------------------------------
+
+While we want all dependencies explicitly pinned in order to benefit from
+consistent testing and development environments, it isn't acceptable to leave
+these versions untouched for long stretches of time.  Packages we depend on
+routinely release new versions to address security issues, fix bugs, and add
+new features.  While we don't necessarily need to update our repositories
+every time a new dependency version is released, we do want to keep them
+current enough that upgrading a single package to fix a known issue doesn't
+require suddenly adapting to a few years' worth of API changes that we didn't
+pay attention to.
+
+Each Open edX repository should have the following:
+
+* An ``upgrade`` make target as described above, to update the pinned versions
+  of all dependencies (and account for any new or removed indirect
+  dependencies).
+* An automated test suite with reasonably good code coverage, configured to
+  be run on new GitHub pull requests.
+* A service configured to periodically auto-generate a GitHub pull request
+  that tests the output of running ``make upgrade`` (if it results in any
+  changes).  This can either be a service such as `requires.io`_ which tracks
+  new releases of Python package dependencies, or a recurring scheduled job.
+* At least one designated maintainer who receives notifications of the
+  generated pull requests and will merge or fix them as needed.  This
+  maintainer should scan the changelog for each upgraded package to look for
+  changes that merit closer inspection; services like `requires.io`_ and
+  `AllMyChanges.com`_ can make this easier.
+
+.. _requires.io: https://requires.io/
+.. _AllMyChanges.com: https://allmychanges.com/
```

Comment:
Not sure why, but AllMyChanges doesn't know about the last five releases of coverage.py: https://allmychanges.com/p/python/coverage/

Reply:
It looks like a number of packages haven't been getting updated for a year or so, the issue was reported 6 days ago: https://github.com/AllMyChanges/allmychanges.com/issues/65 .  Do you think it's still worth mentioning here, or pull until that gets resolved?  It's missing the last several Django releases also.  [requires.io](https://requires.io/github/edx/edx-platform/requirements/?branch=master) is current on these, but missing changelog links for several packages that are included on AllMyChanges.com.

Reply:
Maybe just a sentence like, "At the time of writing, these services were still in development. Carefully evaluate which is currently reliable."

Reply:
interesting. does requires.io automatically pull in github's release history? From what I can tell it looks like it needs a CHANGES.txt file or similar.

Reply:
These services parse CHANGES.txt (or some other file).  GitHub histories are not good sources of changes, they are too noisy.

Reply:
releases are not noisy at all in general - https://github.com/jgm/pandoc/releases, https://github.com/tqdm/tqdm/releases, etc...

Reply:
Oh, sorry, i misread your comment as "commit history."  I don't know if they use GitHub releases.  Frankly, I hope they don't.  A text file in the repo is a low-tech way to make the information available to everyone, regardless of how the source is stored or transmitted.  Unless I've misunderstood something, GitHub releases are a proprietary feature that separates information from the source tree.

Reply:
I agree changes.txt should take precedence, but where none is available it would be nice to fallback to things like github release notes. It's convenient to use github's release system to link to issues, upload binaries etc. anyway this is probably a discussion not meant for here. I was only asking if anyone knows of good crawlers of https://api.github.com/repos/{REPO}/releases, not if this is a good idea :P

Reply:
The requires.io changelog parsing, at least as of when it was first introduced, is summarized in [this blog post](http://shiningpanda.com/introducing-changelogs.html).  At least at the time, it only looked in PyPI and for appropriately named files in the repo, not at the GitHub releases list or commit history.

There's also the changelog parser for Gemnasium at https://github.com/tech-angels/vandamme , it seems somewhat more comprehensive but I haven't looked specifically to see what it scrapes from GitHub.